stdlib/SparseArrays/src/linalg.jl

```diff
@@ -203,6 +203,28 @@ function spmatmul(A::SparseMatrixCSC{Tv,Ti}, B::SparseMatrixCSC{Tv,Ti};
     return C
 end
 
+# Frobenius inner product: trace(A†B)
+function vecdot(A::SparseMatrixCSC{T1,S1},B::SparseMatrixCSC{T2,S2}) where {T1,T2,S1,S2}
+    m, n = size(A)
+    size(B) == (m,n) || throw(DimensionMismatch("matrices must have the same dimensions"))
+    r = zero(promote_type(T1,T2))
+    @inbounds for j = 1:n
+        for i1 = A.colptr[j]:A.colptr[j+1]-1
+            ra = A.rowval[i1]
+            for i2 = B.colptr[j]:B.colptr[j+1]-1
```

Comment:
you shouldn't need to restart this iterator

Reply:
ah you are right, I'll improve the algorithm

Reply:
it's not the prettiest, but you could do something like:
```julia
ia = ia_nxt; ib = ib_nxt
ia_nxt = A.colptr[j+1]; ib_nxt = B.colptr[j+1]

if ia < ia_nxt && ib <= ib_nxt
    ra = A.rowval[ia];  rb = B.rowval[ib]
    while true
        if ra < rb
            ia += 1
            ia < ia_nxt || break
            ra = A.rowval[ia];
        elseif ra > rb
            ib += 1
            ib < ib_nxt || break
            rb = B.rowval[ib]
        else # ra == rb
            r += A.nzval[ia]' * B.nzval[ib]
            ia += 1; ib += 1
            ia < ia_nxt && ib < ib_nxt || break
            ra = A.rowval[ia]; rb = B.rowval[ib]
        end
    end
end
```

Reply:
I compared yours (which has a small typo on the first line) with one where I just updated the lower bound of the loop and yours was faster